docs/dev/devel-env.rst

```diff
@@ -286,12 +294,157 @@ Also, run the **integration tests** for ListenBrainz.
+-----------------------------
+
+If it's your first time initializing the containers, format the ``namenode`` container
+using the given command.
+
+.. code-block:: bash
+
+    docker-compose -f docker/docker-compose.spark.yml -p listenbrainzspark run --rm hadoop-master hdfs namenode -format -nonInteractive -force
+
+This will format the ``namenode`` and link the ``namenode`` and ``datanode`` to a single ClusterID. 
+
+Remember that you shouldn't execute this command after running it for the first time. If run again it will
+format the ``namenode`` again and its link with the ``datanode`` will be lost, as both of them will have different
+ClusterID's.
+
+Your development environment is now ready. Now, let's actually see ListenBrainz Spark
+load locally!
+
+
+Run the magic script
+--------------------
+
+Start your development environment by executing ``develop.sh spark up``. Now, it 
+will work as expected.
+
+.. code-block:: bash
+
+    ./develop.sh spark up
+
+Now, you are all set to begin making changes and seeing them in real-time inside
+of your development environment!
+
+Once you are done with your work, shut down the containers using the following command. 
+
+.. code-block:: bash
+
+    ./develop.sh spark down
+
+
+Test your changes with unit tests
 ---------------------------------
 
-* What to do if getting an error while running './develop.sh build' command, 'ERROR: Couldn't connect to Docker daemon at http+docker://localhost - is it running?' ?
+Unit tests are an important part of ListenBrainz Spark. It helps make it easier for
+developers to test changes and help prevent easily avoidable mistakes later on.
+Before committing new code or making a pull request, run the unit tests on your
+code.
+
+.. code-block:: bash
+
+   ./spark_test.sh
+
+This builds and runs the containers needed for the tests. This script configures
+test-specific data volumes so that test data is isolated from your development
+data.
+
+Also, run the **integration tests** for ListenBrainz Spark.
 
--  You need to add the user to the docker group by entering the following command in the terminal :
-  `sudo usermod -aG docker $USER`. After this command, restart the computer and then again run the './develop.sh build'.
+.. code-block:: bash
+
+   ./integration-test.sh
+
+When the tests complete, you will see if your changes are valid or not. These tests
+are a helpful way to validate new changes without a lot of work.
 
 
+FAQ's
+=====
+
+**What to do if getting an error while running ``./develop.sh build`` command, 'ERROR: Couldn't connect to Docker daemon at http+docker://localhost - is it running?'?**
+
+- You need to add the user to the docker group by entering the following command in the terminal:
+  
+  .. code-block:: bash
+
+    sudo usermod -aG docker $USER
+
+- After this command, restart the computer and then again run ``./develop.sh build.``
+|
+**How to resolve 'ERROR: NameNode not formatted' which arises on running ``develop.sh spark up``?**
```

Comment:
if namenode is formatted, datanode should be formatted too.
I think we can skip this for now because I am working on adding this to Hacking.md, with detailed commands. Later we can add it in the FAQ with a warning about what formatting can do to the data and then link HAcking.md for the command to format.

Reply:
Okay I will remove thisquestion  till then.